source/state_representation/include/state_representation/trajectory/TrajectoryBase.hpp

```diff
@@ -0,0 +1,259 @@
+  std::deque<std::chrono::nanoseconds> times_;
+
+public:
+  /**
+   * @brief Empty constructor
+   */
+  explicit TrajectoryBase();
+
+  /**
+   * @brief Constructor with name provided
+   * @brief name the name of the state
+   */
+  explicit TrajectoryBase(const std::string& name);
+
+  /**
+   * @brief Get attribute list of trajectory points
+   */
+  const std::deque<TrajectoryT>& get_points() const;
+
+  /**
+   * @brief Get the trajectory point at given index
+   * @param index the index
+   */
+  const TrajectoryT& get_point(unsigned int index) const;
+
+  /**
+   * @brief Reset trajectory
+   */
+  virtual void reset();
+
+  /**
+   * @brief Delete last point and corresponding time from trajectory
+   */
+  virtual void delete_point();
+
+  /**
+   * @brief Clear trajectory
+   */
+  virtual void clear();
+
+  /**
+   * @brief Get attribute list of trajectory times
+   */
+  const std::deque<std::chrono::nanoseconds>& get_times() const;
+
+  /**
+   * @brief Get attribute number of point in trajectory
+   */
+  int get_size() const;
+
+protected:
+  /**
+   * @brief Add new point and corresponding time to trajectory
+   */
+  template<typename DurationT>
+  void add_point(const TrajectoryT& new_point, const std::chrono::duration<int64_t, DurationT>& new_time);
+
+  /**
+   * @brief Insert new point and corresponding time to trajectory between two
+   * already existing points
+   */
+  template<typename DurationT>
+  void insert_point(const TrajectoryT& new_point, const std::chrono::duration<int64_t, DurationT>& new_time, int pos);
+
+  /**
+   * @brief Get the trajectory point at given index
+   * @param index the index
+   */
+  TrajectoryT& get_point(unsigned int index);
+
+  /**
+   * @brief Set the trajectory point at given index
+   * @param index the index
+   * @param point the new point
+   * @param new_time the new time
+   * @return Success of the operation
+   */
+  template<typename DurationT>
+  bool
+  set_point(unsigned int index, const TrajectoryT& point, const std::chrono::duration<int64_t, DurationT>& new_time);
+
+  /**
+   * @brief Set the trajectory point at given index
+   * @param points vector of new points
+   * @param new_time vector of new times
+   * @return Success of the operation
+   */
+  template<typename DurationT>
+  bool set_points(
+      const std::vector<TrajectoryT>& points, const std::vector<std::chrono::duration<int64_t, DurationT>>& new_times);
+
+  /**
+   * @brief Operator overload for returning a single trajectory point and
+   * corresponding time
+   */
+  const std::pair<TrajectoryT, std::chrono::nanoseconds> operator[](unsigned int idx) const;
+
+  /**
+   * @brief Operator overload for returning a single trajectory point and
+   * corresponding time
+   */
+  std::pair<TrajectoryT, std::chrono::nanoseconds> operator[](unsigned int idx);
+};
+
+template<typename TrajectoryT>
+inline TrajectoryBase<TrajectoryT>::TrajectoryBase() : State() {
+  this->set_type(StateType::NONE);
```

Comment:
I don't think this is necessary, as long as the constructor is not accessible and we do it correctly in the derived classes